lib/plug_datadog_stats.ex

```diff
@@ -1,43 +1,69 @@
 defmodule PlugDatadogStats do
   @behaviour Plug
-  import Plug.Conn, only: [register_before_send: 2]
+  import Plug.Conn, only: [register_before_send: 2, put_private: 3]
   require Logger
 
-  def init(_) do
-    %{
-      histogram: Application.get_env(:plug_datadog_stats, :histogram_name, "resp_time"),
-      count: Application.get_env(:plug_datadog_stats, :count_name, "resp_count"),
-    }
-  end
+  @base_opts %{
+    histogram: Application.get_env(:plug_datadog_stats, :histogram_name, "resp_time"),
+    count: Application.get_env(:plug_datadog_stats, :count_name, "resp_count"),
+    path_normalize_fn: {__MODULE__, :path_override_passthrough, []}
+  }
 
-  def call(conn, names) do
-    req_start_time = :os.timestamp
+  def init(opts \\ []) do
+    opts
+    |> Enum.into(@base_opts)
+    |> parse_options()
+  end
 
-    register_before_send conn, fn conn ->
-      tags = tags_for_conn(conn)
+  def call(conn, opts) do
+    req_start_time = :os.timestamp()
 
+    register_before_send(conn, fn conn ->
       # This will log response time in microseconds
-      req_end_time = :os.timestamp
+      req_end_time = :os.timestamp()
       duration = :timer.now_diff(req_end_time, req_start_time)
 
-      Logger.debug("PlugDatadogStats: #{duration}µs #{names.histogram}/#{names.count} #{inspect tags}")
-      ExStatsD.histogram(duration, names.histogram, tags: tags)
-      ExStatsD.increment(names.count, tags: tags)
+      tags = tags_for_conn(conn, opts.path_normalize_fn)
 
-      conn
-    end
+      Logger.debug(
+        "PlugDatadogStats: #{duration}µs #{opts.histogram}/#{opts.count} #{inspect(tags)}"
+      )
+
+      ExStatsD.histogram(duration, opts.histogram, tags: tags)
+      ExStatsD.increment(opts.count, tags: tags)
+
+      # Tags are added to the conn for testing and observability purposes
+      put_private(conn, :plug_datadog_stats_tags, tags)
+    end)
   end
 
-  defp tags_for_conn(conn) do
+  defp parse_options(%{path_normalize_fn: {_m, _f, _a}} = opts) do
+    opts
+  end
+
+  defp parse_options(%{path_normalize_fn: _bad}) do
+    raise """
```

Comment:
Question largely stemming from my lack of knowledge - is this message going to get logged for the application or would the application using this need to implement some logic to log the error?

Reply:
This method is called at compile time so this is what you'd see locally or in CI if you passed the `path_normalize_fn` with a bad value when declaring the plug.

The other option is just not to have this function clause, but then you'd get a less helpful `FunctionClauseError` in CI.

You're right that if this were a runtime message we'd see it in the error logs when the process crashes due to an unhandled `raise` or we'd have to rescue it and log it out.